WIP: Shadow

curran opened the pull request:
Creating a PR from this branch so we can discuss.

seemantk committed:
I think this sort of structure could yield some or all:
	* responsiveness
	* zoom/pan via viewBox
	* single renderer
	* logical groupings by presentation
	* (logical groupings by behavior in the code)
	* colors purely via css
viewBox

seemantk committed:
Conceptual attributes (ie, not literal and not valid) are in parens

seemantk committed:
Use javascript filenames as the id names for each component.

seemantk committed:
space

seemantk committed:
Add a "notes" class.  And shorten a line to be closer to 80 chars.

seemantk committed:
Add discarded alternative (with reason for discard).

seemantk committed:
Clean up comment documentation.

seemantk committed:
better phrasing

seemantk committed:
Add placeholders for the different ribbon types.
Visiblity/Display/Opacity can toggle between the two.  Added this to
both alternatives.

seemantk committed:
Use d3.queue() so that multiple files can potentially be read in.

seemantk committed:
Disconnect and disable the notesnav and notesbook modules.

seemantk committed:
Instead of the notesbook and notesnav, instantiate a notescanvas.

seemantk committed:
Add a slugify() function to convert strings to slugs.

seemantk committed:
Reboot notescanvas. Use @curran's object methodology (svg() and data()
API methods).  This is a skeleton object that creates the skeleton
shadow svg.

seemantk committed:
Add an object to create a strip of notes for one voice.  Current
functionality is to complete the svg skeleton.

seemantk committed:
Add .x() and .y() API methods, because notescanvas will call those with
global scales.

seemantk committed:
Include the new score.js

seemantk committed:
Restore missing brace.  oops.

seemantk committed:
Let the generator be d3.entries() from the get-go.  Call the module,
passing in the x and y scales.

seemantk committed:
Let the generator itself be a plain object.  d3.entries() it only for
the selection purposes.

seemantk committed:
Since I can't get the update selection to work, make a workaround
(hoping @curran can demystify this).

seemantk committed:
Create a container on the DOM to test the ```<use>``` functionality now.

seemantk committed:
Small cleanups -- vars rearranged location.  The svg check is now a
check for !svg with a quick bailout.

seemantk committed:
Minimize large fork.

seemantk committed:
This is the proposal for the shadow version.

seemantk committed:
Add mouse interaction test, to see if it will work when instantiated on
the DOM.

seemantk committed:
construct the notesbook and notescanvas with the shadow DOM pieces.

seemantk committed:
Add colorbrewer colors here.

seemantk committed:
Add scaling/responsive stuff.  The viewBox dimensions are determined
from the dataset's extents, with an assumption of physical size for a
note of one time unit.

seemantk committed:
The viewBox should vertically correspond to the y scale.

seemantk committed:
Assign different heights for the nav and main ```<svg>```s

curran commented on proposed-skeleton.svg:
```diff
@@ -0,0 +1,50 @@
+    <g class="score voice5 bassus"></g>
+  </symbol><!--/#score-->
+  <symbol id="ribbon" viewBox="0 0 max(time + duration) mymaxpitch">
+    <g class="ribbon voice0 superius"></g>
```
Probably for voices we should leave it as data-driven, as there could possibly be a voice name we did not anticipate here.

seemantk replied:
Nice catch, @curran!

curran commented on proposed-skeleton.svg:
```diff
@@ -0,0 +1,50 @@
+   <g class="brush"></g>
+</svg><!--/#notesnav-->
+
+<!-- Shadow DOM (not directly rendered) -->
```
Do you mean really using the [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Shadow_DOM), where we'll use methods like [Element.attachShadow()](https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow) and [Element.shadowRoot](https://developer.mozilla.org/en-US/docs/Web/API/Element/shadowRoot) or is this more metaphorical?

curran commented on proposed-skeleton.svg:
```diff
@@ -0,0 +1,50 @@
+    <g class="reflines voice4 altus"></g>
+    <g class="reflines voice5 bassus"></g>
+  </symbol><!--/#reflines-->
+</svg><!--.shadow-->
```
Overall I think it's a great idea to have a fixed skeleton for things that we know will never change.

curran commented on proposed-skeleton.svg:
````diff
@@ -48,3 +48,44 @@
+
+<!-- This would be cleaner to generate and process, but it doesn't really
+      allow for transitions to be made universally.  Each instance that
+      ```<use>```s the symbols has to manually append the voices together, and
````
I don't understand the need for `<use>` here, maybe let's discuss.

curran commented on proposed-skeleton.svg:
````diff
@@ -19,9 +19,14 @@
+      The combine/separate transitions can apply to this svg and get rendered
+      to every ```<use>``` automatically.
+
+      Right?
````
I'm not fully following the logic. I'm not sure we need to pull in the whole Shadow DOM technology for our use case.

curran commented on js/notescanvas.js:
```diff
@@ -343,6 +270,20 @@ function NotesCanvas() {
     /*
     ** API (Setter ONLY) Functions
     */
+    my.svg = function(_) {
```
Nice. It would be cool to unify the codebase to use this approach, and have all the `my()` methods callable multiple times with changing data.

curran commented on js/score.js:
```diff
@@ -20,26 +20,35 @@ function Score() {
-
-      var rects = selection.selectAll("rect")
+      var rect = selection.selectAll("rect")
               .data(function(d) { return d; })
```
This bit is confusing, as this function depends on how the selection was constructed before it was passed in. Would be great to use the object methodology here as well, so we can construct the selection inside the function.

curran replied:
Tracked down where the selection is getting constructed, in `notescanvas`:

``` js
    function generate(g) {
        var self = d3.select(this);
        self.selectAll("." + g.key)
            .data(data.partnames, function(d) { return d; })
          .enter().append("g")
            .attr("class", function(d, i) {
                return [g.key, slugify(d), ("voice" + i)].join(' ');
              })
          .each(function(d) {
              var that = d3.select(this);
              console.log(that, d, g);
              that
                  .datum(data.notes.get(d))
                  .call(g.value.x(x).y(y))
              ;
            })
        ;
    } // generate()
```

I think the issue might be we're using the Towards Reusable Charts pattern here, but not using `each` inside `my`.

curran replied:
Or maybe we're not getting `enter` because `datum()` was used, and we never created a new selection that uses `data()`. I believe you only get `enter()` when using `data()`, but not with `datum()`.

curran commented on js/baton.js:
```diff
@@ -60,7 +60,19 @@ function parseJSON(proll) {
+    ;
+    notesnav
+      .append("use")
+        .attr("xlink:href", "#ribbon")
```
Interesting. So we can manage the DOM separately, and it gets injected onto the page here via `use`?

seemantk replied:
You're correct!  this isn't the deep/full shadow DOM. Things like `<symbol>` and `<defs>` are never directly rendered, for example: they only get rendered via `<use>`.

curran reviewed (commented):
Appreciate your efforts here @seemantk . Looking forward to see it all come together! Will try investigating further into those selection issues.

seemantk committed:
Rename, as this default name is appropriate for this branch.

seemantk committed:
See if we can limit the height of this container element to the viewport
size.

seemantk committed:
Remove the mouse interactions for the moment.

seemantk committed:
Add one to the y-range (a la @curran).  Set the viewBox to match the x
and y scales.

seemantk committed:
Fix the preserveAspectRatio.

seemantk committed:
Increase the dimensions again for now to fill up the entire viewport.

seemantk committed:
Shorten the code, since notesbook and notesnav are both being built
identically.

seemantk committed:
Calculate the default viewbox

seemantk committed:
more DRY

seemantk committed:
replace the snap function with update(), using an argument.

seemantk committed:
Switch to using update(false) instead of snap()

seemantk committed:
better abbrvtn for transition.

seemantk committed:
Refactor all of notesbook and notesnav.  add a viewBox attribute to
notesBook to magnify and show a portion of the notestrip.

seemantk committed:
Undo that append thing, that was silly.  Also, for the container svg's
viewbox in notesbook, we only need the shadow svg's width and height.
Now, we have a responsive situation. /cc @curran.

seemantk commented:
@curran, this took me a few tries but I think I have something that is responsive.  Please note, that I have only the barest of SVG yet (the notes only, and they're static still), so this is just for you to see if the resizing responsiveness is intuitive for you.

seemantk commented:
(It isn't intuitive to me, as it shows more when smaller and less when bigger.  The other way round feels like it would be right.

seemantk committed:
Add the full 10category colorbrewer2 set.

seemantk commented:
![first-brush](https://cloud.githubusercontent.com/assets/589271/20128298/159b6726-a5fc-11e6-9c86-d9ed94824e8e.png)

Brush-work in progress.
